proposals/0000-record-dot-syntax.md

```diff
@@ -0,0 +1,316 @@
+
+## Motivation
+
+In almost every programming language we write `a.b` to mean the `b` field of the `a` record expression. In Haskell that becomes `b a`, and even then, only works if there is only one `b` in scope. Haskell programmers have struggled with this weakness, variously putting each record in a separate module and using qualified imports, or prefixing record fields with the type name. We propose bringing `a.b` to Haskell, which works regardless of how many `b` fields are in scope. Here's a simple example of what is on offer:
+
+```haskell
+{-# LANGUAGE RecordDotSyntax #-}
+
+data Company = Company {name :: String, owner :: Person}
+data Person = Person {name :: String, age :: Int}
+
+display :: Company -> String
+display c = c.name ++ " is run by " ++ c.owner.name
+
+nameAfterOwner :: Company -> Company
+nameAfterOwner c = c{name = c.owner.name ++ "'s Company"}
+```
+
+We declare two records both having `name` as a field label. The user may then write `c.name` and `c.owner.name` to access those fields. We can also write `c{name = x}` as a record update, which works even though `name` is no longer unique. Under the hood, we make use of `getField` and `setField` from [the `HasField` proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0158-record-set-field.rst).
+
+An implementation of this proposal has been battle tested and hardened over 18 months in the enterprise environment as part of [Digital Asset](https://digitalasset.com/)'s [DAML](https://daml.com/) smart contract language (a Haskell derivative utilizing GHC in its implementation), and also in a [Haskell preprocessor and a GHC plugin](https://github.com/ndmitchell/record-dot-preprocessor/). When initially considering Haskell as a basis for DAML, the inadequacy of records was considered the most severe problem, and without devising the scheme presented here, we wouldn't be using Haskell. The feature enjoys universal popularity with users.
+
+## Proposed Change Specification
+
+For the specification we focus on the changes to the parsing rules, and the desugaring, with the belief the type checking and renamer changes required are an unambiguous consequences of those.
+
+### `RecordDotSyntax` language extension
+
+This change adds a new language extension `RecordDotSyntax`. In the event the language extension is enabled:
+
+| Expression | Equivalent |
+| -- | -- |
+| `e.lbl` | `getField @"lbl" e` the `.` cannot have whitespace either before or after |
+| `e{lbl = val}` | `setField @"lbl" e val` |
+| `.lbl` | `(\x -> x.lbl)` the `.` cannot have whitespace after |
+| `e{lbl1.lbl2 = val}` | `e{lbl1 = (e.lbl1){lbl2 = val}}` performing a nested update |
+
+The above forms combine to provide these identities:
+
+| Expression | Identity
```

Comment:
I don't know what you mean by identities here. But, if some of these could be considered simple syntactic rules (_e.g._ `e.lbl1.lbl2` parses as `(e.lbl1).lbl2`), I don't really see the difference between the contents of this table and that of the previous one. Maybe merge the tables?

Reply:
I interpreted the text to mean that the identities logically follow from the definitions. (That is, if we just implement the definitions, we get the identities "for free".) But I don't think this is actually the case. I don't see how we get identity (2) from the above table, for example.

Reply:
@aspiwack - thanks for all your remarks, I believe we've addressed them in the document.